addons/spreadsheet_dashboard/controllers/share.py

```diff
@@ -0,0 +1,75 @@
+from werkzeug.exceptions import Forbidden
+
+from odoo import http, _
+from odoo.http import request
+
+class DashboardShareRoute(http.Controller):
+    @http.route(['/dashboard/share/<int:share_id>/<token>'], type='http', auth='public')
+    def share_portal(self, share_id=None, token=None, **kw):
+        share = request.env["spreadsheet.dashboard.share"].sudo().browse(share_id).exists()
+        if share:
+            dashboard = share._get_dashboard_and_check_access(token)
+            if dashboard:
+                return self.open_dashboard(share, token)
+            else:
+                raise Forbidden(_("You don't have access to this dashboard"))
+        return request.not_found()
+
+    @http.route(["/dashboard/download/<int:share_id>/<access_token>/<int:dashboard_id>"],
+                type='http', auth='public')
+    def download_one(self, dashboard_id=None, access_token=None, share_id=None):
+        dashboard = request.env["spreadsheet.dashboard"].sudo().browse(dashboard_id).exists()
+        if dashboard:
+            share = request.env["spreadsheet.dashboard.share"].sudo().browse(share_id)
+            available_dashboard = share._get_dashboard_and_check_access(access_token)
+            if not available_dashboard or dashboard != available_dashboard:
+                raise Forbidden(_("You don't have access to this dashboard"))
+            stream = request.env["ir.binary"]._get_stream_from(
+                share, "excel_export", filename=dashboard.name
+            )
+            return stream.get_response()
+        return request.not_found()
+
+    @http.route(
+        ["/dashboard/data/<int:dashboard_share_id>/<token>"],
+        type="http",
+        auth="public",
+        methods=["GET"],
+    )
+    def get_shared_dashboard_data(self, dashboard_share_id, token):
+        share = (
+            request.env["spreadsheet.dashboard.share"]
+            .sudo()
+            .browse(dashboard_share_id)
+            .exists()
+        )
+        if not share:
+            raise request.not_found()
+
+        dashboard = share._get_dashboard_and_check_access(token)
+        if not dashboard:
+            raise Forbidden(_("You don't have access to this dashboard"))
```

Comment:
you can actually move the repetitive "if not dashboard: raise" inside of `_get_dashboard_and_check_access`.
See the other comment ;)